drivers/console/uart_console.c

```diff
@@ -446,13 +446,13 @@ static void uart_console_isr(const struct device *unused, void *user_data)
 	ARG_UNUSED(user_data);
 	static uint8_t last_char = '\0';
 
-	while (uart_irq_update(uart_console_dev) &&
-	       uart_irq_is_pending(uart_console_dev)) {
+	while (uart_irq_update(uart_console_dev) > 0 &&
+	       uart_irq_is_pending(uart_console_dev) > 0) {
 		static struct console_input *cmd;
 		uint8_t byte;
 		int rx;
 
-		if (!uart_irq_rx_ready(uart_console_dev)) {
+		if (uart_irq_rx_ready(uart_console_dev) == 0) {
```

Comment:
This doesn't look right -- if the rx_ready function returns failure, the isr should probably return. Probably need to check for three cases -- not ready (continue), failure (return), ready (read a byte).

Reply:
I can test it. The current code works though as well.

Note: The logic here isn't changed from before

Reply:
Yup, I realize you haven't changed how it works, but it sure looks like it was broken before :-)

Reply:
@dcpleung - I'm curious to know what you think in this case. The proposal is to change this conditional from
```cpp
		if (uart_irq_rx_ready(uart_console_dev) == 0
			/* formerly !uart_irq_rx_ready(uart_console_dev) */) {
  			continue;
		}
```
to this 
```cpp
		if (uart_irq_rx_ready(uart_console_dev) <= 0) {
			return;
		}
```

Twister / qemu tests work for me here with the latter, I'm just not 100% sure if it will work exactly the same on every target supported by Zephyr. And that logic goes all the way back to when the driver was first added.

Reply:
The old code looks like it was polling waiting for RX ready; you're proposing to have it simply bail if there's nothing pending instead of spinning. That seems like a pretty significant semantic change...

Reply:
> The old code looks like it was polling waiting for RX ready; you're proposing to have it simply bail if there's nothing pending instead of spinning. That seems like a pretty significant semantic change...

Ah, sorry - I misread your earlier message.

Reply:
@keith-packard - done!